File: oneflow/python/eager/opkernel_object.py
"""
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
from __future__ import absolute_import

import oneflow.python.eager.symbol_storage as symbol_storage
import oneflow.python.eager.symbol as symbol_util
import oneflow.python.framework.c_api_util as c_api_util
import oneflow_api


class OpKernelObject(oneflow_api.Object):
    def __init__(self, object_id, op_conf, release):
        oneflow_api.Object.__init__(self, object_id, _GetOpParallelSymbol(op_conf))
        self.op_conf_ = op_conf
        self.scope_symbol_ = _GetScopeSymbol(op_conf)
        self.release_ = []
        if release is not None:
            self.release_.append(release)

    @property
    def op_conf(self):
        return self.op_conf_

    @property
    def scope_symbol(self):
        return self.scope_symbol_

    def __del__(self):
        for release in self.release_:
            release(self)
        self.release_ = []


def _GetScopeSymbol(op_conf):
    assert op_conf.HasField("scope_symbol_id")
    return oneflow_api.GetScopeSymbol(op_conf.scope_symbol_id)


def _GetOpParallelSymbol(op_conf):
    assert op_conf.HasField("scope_symbol_id")
    symbol_id = c_api_util.GetOpParallelSymbolId(op_conf)
    return oneflow_api.GetPlacementSymbol(symbol_id)

File: oneflow/core/framework/compatible_py/opkernel_object.cpp
/*
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
#include "oneflow/core/framework/opkernel_object.h"
#include "oneflow/core/framework/symbol_storage_util.h"

namespace oneflow {

namespace compatible_py {

namespace {

std::shared_ptr<Scope> GetScopeSymbol(const std::shared_ptr<cfg::OperatorConf>& op_conf) {
  CHECK(op_conf->has_scope_symbol_id());
  return CHECK_JUST(GetSymbol<cfg::ScopeProto, Scope>(op_conf->scope_symbol_id()));
}

std::shared_ptr<ParallelDesc> GetOpParallelSymbol(
    const std::shared_ptr<cfg::OperatorConf>& op_conf) {
  CHECK(op_conf->has_scope_symbol_id());
  const auto& scope = Global<symbol::Storage<Scope>>::Get()->Get(op_conf->scope_symbol_id());
  OperatorConf pb_op_conf;
  op_conf->ToProto(&pb_op_conf);
  int64_t parallel_desc_symbol_id = CHECK_JUST(scope.GetParallelDescSymbolId(pb_op_conf));
  return CHECK_JUST(GetSymbol<cfg::ParallelConf, ParallelDesc>(parallel_desc_symbol_id));
}

}  // namespace

OpKernelObject::OpKernelObject(int64_t object_id, const std::shared_ptr<cfg::OperatorConf>& op_conf,
                               const std::function<void(Object*)>& release)
    : Object(object_id, GetOpParallelSymbol(op_conf)),
      op_conf_(op_conf),
      scope_symbol_(GetScopeSymbol(op_conf)) {
  release_.emplace_back(release);
}

void OpKernelObject::ForceReleaseAll() {
  for (const auto& release : release_) { release(this); }
}

}  // namespace compatible_py

}  // namespace oneflow
